apptest/tests/replication_test.go

```diff
@@ -0,0 +1,472 @@
+	// that the remaining vmstorage nodes must still be able to provide full
+	// response.
+	for i := range replicationFactor - 1 {
+		tc.StopApp(c.vmstorages[i].Name())
+	}
+	assertSeries(c.vmselect, mustReturnPartialResponse)
+	assertSeries(c.vmselectRF, mustReturnFullResponse)
+
+	// Stop one more vmstorage. At this point the remaining vmstorage nodes are
+	// not enough to provide the full dataset. Therefore both vmselects must
+	// return partial response.
+	tc.StopApp(c.vmstorages[replicationFactor].Name())
+	assertSeries(c.vmselect, mustReturnPartialResponse)
+	assertSeries(c.vmselectRF, mustReturnPartialResponse)
+}
+
+// TestClusterReplication_SkipSlowReplicas checks that vmselect skips the
+// results from slower replicas if the results that have been received from
+// other replicas are enough to construct full response.
+//
+// By default, even if a vmselect knows about the vmstorage replication (via
+// -replicationFactor flag) it will still wait for results from all the
+// vmstorage nodes. A vmselect can be configured to skip slow replicas
+// using -search.skipSlowReplicas flag.
+//
+// Say a vmselect points to N vmstorage nodes and its -replicationFactor is R.
+// Then only R nodes out of N will contain the searched data, while N-R node
+// will not contain it. Therefore the vmselect must receive responses from at
+// least N-R+1 nodes to construct the full response. The responses from the rest
+// of the nodes (R-1) can be skipped.
+//
+// See: https://docs.victoriametrics.com/cluster-victoriametrics/#replication-and-data-safety
+func TestClusterReplication_SkipSlowReplicas(t *testing.T) {
+	tc := at.NewTestCase(t)
+	defer tc.Stop()
+
+	const replicationFactor = 2
+	c := newClusterWithReplication(tc, replicationFactor)
+
+	// Insert data.
+
+	const numRecs = 1000
+	recs := make([]string, numRecs)
+	wantSeries := &at.PrometheusAPIV1SeriesResponse{
+		Status: "success",
+		Data:   make([]map[string]string, numRecs),
+	}
+	for i := range numRecs {
+		name := fmt.Sprintf("metric_%d", i)
+		recs[i] = fmt.Sprintf("%s %d", name, rand.IntN(1000))
+		wantSeries.Data[i] = map[string]string{"__name__": name}
+	}
+	wantSeries.Sort()
+	c.vminsert.PrometheusAPIV1ImportPrometheus(t, recs, at.QueryOpts{})
+	tc.ForceFlush(c.vmstorages...)
+
+	// Verify skipping slow replicas by counting the number of skipSlowReplicas
+	// messages in request trace.
+
+	assertSeries := func(app *at.Vmselect, want int) {
+		t.Helper()
+		tc.Assert(&at.AssertOptions{
+			Msg: "unexpected /api/v1/series response",
+			Got: func() any {
+				return app.PrometheusAPIV1Series(t, `{__name__=~".*"}`, at.QueryOpts{}).Sort()
+			},
+			Want: wantSeries,
+		})
+
+		res := app.PrometheusAPIV1Series(t, `{__name__=~".*"}`, at.QueryOpts{Trace: "1"})
+		got := res.Trace.Contains("cancel request because -search.skipSlowReplicas is set and every group returned the needed number of responses according to replicationFactor")
```

Comment:
This check doesn't look reliable to me. What if trace message will be changed? 

Is it possible to use some metric instead?

Reply:
It is reliable unless the trace message is changed in the code. If the message is changed, then the test must be adjusted accordingly.